File: GP_ITI_ParkingAssis/packages/apps/buzzertest/app/src/main/java/com/example/buzzertest/MainActivity.java
package com.example.buzzertest;

import android.os.Bundle;
import android.os.Handler;
import android.os.IBinder;
import android.os.Looper;
import android.os.RemoteException;
import android.util.Log;
import android.widget.Button;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.luxoft.buzzer.IBuzzer;

import java.lang.reflect.Method;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "BUZZER_TEST";
    private Button btn0, btn1, btn2, btn3, btn4;
    private IBuzzer iBuzzer;
    private int currentLevel = -1; // -1 indicates no active level
    private boolean isRunning = false;
    private Handler handler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        // Initialize buttons
        btn0 = findViewById(R.id.btn_0);
        btn1 = findViewById(R.id.btn_1);
        btn2 = findViewById(R.id.btn_2);
        btn3 = findViewById(R.id.btn_3);
        btn4 = findViewById(R.id.btn_4);

        handler = new Handler(Looper.getMainLooper());

        // Bind to the IBuzzer service
        try {
            Class<?> serviceManagerClass = Class.forName("android.os.ServiceManager");
            Method getServiceMethod = serviceManagerClass.getMethod("getService", String.class);
            Object service = getServiceMethod.invoke(null, "com.luxoft.buzzer.IBuzzer/default");

            if (service != null) {
                IBinder binder = (IBinder) service;
                iBuzzer = IBuzzer.Stub.asInterface(binder);
            } else {
                Toast.makeText(this, "Buzzer service not found", Toast.LENGTH_SHORT).show();
                Log.e(TAG, "Buzzer service not available.");
                return;
            }
        } catch (Exception e) {
            Toast.makeText(this, "Failed to bind buzzer service: " + e.getMessage(), Toast.LENGTH_LONG).show();
            Log.e(TAG, "Error binding buzzer service", e);
            return;
        }

        // Set button click listeners
        btn0.setOnClickListener(v -> stopBuzzer());
        btn1.setOnClickListener(v -> startBuzzer(1));
        btn2.setOnClickListener(v -> startBuzzer(2));
        btn3.setOnClickListener(v -> startBuzzer(3));
        btn4.setOnClickListener(v -> startBuzzer(4));
    }

    private void startBuzzer(int level) {
        if (currentLevel != level) {
            currentLevel = level;
            isRunning = true;
            Toast.makeText(this, "Starting buzzer at level " + level, Toast.LENGTH_SHORT).show();
            Log.i(TAG, "Buzzer level " + level + " started.");
            runBuzzer();
        } else {
            Toast.makeText(this, "Buzzer already running at level " + level, Toast.LENGTH_SHORT).show();
        }
    }

    private void stopBuzzer() {
        isRunning = false;
        currentLevel = -1;
        try {
            if (iBuzzer != null) {
                boolean success = iBuzzer.setbuzzerLevel(0); // Stop the buzzer
                if (success) {
                    Toast.makeText(this, "Buzzer stopped", Toast.LENGTH_SHORT).show();
                    Log.i(TAG, "Buzzer stopped.");
                } else {
                    Toast.makeText(this, "Failed to stop buzzer", Toast.LENGTH_SHORT).show();
                }
            }
        } catch (RemoteException e) {
            Toast.makeText(this, "Error stopping buzzer", Toast.LENGTH_SHORT).show();
            Log.e(TAG, "RemoteException while stopping buzzer", e);
        }
    }

    private void runBuzzer() {
        handler.post(new Runnable() {
            @Override
            public void run() {
                if (isRunning && currentLevel != -1) {
                    try {
                        if (iBuzzer != null) {
                            boolean success = iBuzzer.setbuzzerLevel(currentLevel);
                            if (!success) {
                                Log.e(TAG, "Failed to set buzzer level " + currentLevel);
                            }
                        }
                    } catch (RemoteException e) {
                        Log.e(TAG, "RemoteException while running buzzer", e);
                    }

                    // Repeat after a delay if still running
                    handler.postDelayed(this, 1000); // Adjust delay as necessary
                }
            }
        });
    }
}

File: GP_ITI_ParkingAssis/packages/apps/ParkAssistAidlTestTwo/java/com/luxoft/adilultrasonickotlintest/MainActivity.kt
package com.luxoft.adilultrasonickotlintest

import android.os.Bundle
import android.util.Log
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch

class MainActivity : AppCompatActivity() {

    private lateinit var txtSteeringWheelAngle: TextView


    private lateinit var txtUltrasonicOne: TextView
    private lateinit var txtUltrasonicTwo: TextView
    private lateinit var txtUltrasonicThree: TextView
    private lateinit var txtUltrasonicFour: TextView
    private lateinit var txtUltrasonicFive: TextView
    private lateinit var txtUltrasonicSix: TextView

    private lateinit var btnStop: Button
    private lateinit var btnLevel1: Button
    private lateinit var btnLevel2: Button
    private lateinit var btnLevel3: Button
    private lateinit var btnLevel4: Button


    private val repository = ParkingAssistantRepository()
    private val viewModel: ParkingAssistantViewModel by viewModels {
        ParkingAssistantViewModelFactory(repository)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // Initialize TextView references
        txtSteeringWheelAngle = findViewById(R.id.txtSteeringWheelAngle)
        txtUltrasonicOne = findViewById(R.id.txtUltrasonicOne)
        txtUltrasonicTwo = findViewById(R.id.txtUltrasonicTwo)
        txtUltrasonicThree = findViewById(R.id.txtUltrasonicThree)
        txtUltrasonicFour = findViewById(R.id.txtUltrasonicFour)
        txtUltrasonicFive = findViewById(R.id.txtUltrasonicFive)
        txtUltrasonicSix = findViewById(R.id.txtUltrasonicSix)

        btnStop = findViewById(R.id.btn_0)
        btnLevel1 = findViewById(R.id.btn_1)
        btnLevel2 = findViewById(R.id.btn_2)
        btnLevel3 = findViewById(R.id.btn_3)
        btnLevel4 = findViewById(R.id.btn_4)

        lifecycleScope.launch {
            viewModel.steeringWheelAngle.collect { angle ->

                if (angle in 0..180) {
                    txtSteeringWheelAngle.text = "Steering Angle: $angle°"
                }
            }
        }

        lifecycleScope.launch {
            viewModel.currentLevel.collect { level ->
                if (level == -1) {
                    Toast.makeText(this@MainActivity, "Buzzer stopped", Toast.LENGTH_SHORT).show()
                } else {
                    Toast.makeText(this@MainActivity, "Buzzer running at level $level", Toast.LENGTH_SHORT).show()
                }
            }
        }

        btnStop.setOnClickListener {
            viewModel.stopBuzzer()
            Toast.makeText(this, "Stop button clicked", Toast.LENGTH_SHORT).show()
        }

        btnLevel1.setOnClickListener {
            viewModel.startBuzzer(1)
            Toast.makeText(this, "Level 1 button clicked", Toast.LENGTH_SHORT).show()
        }

        btnLevel2.setOnClickListener {
            viewModel.startBuzzer(2)
            Toast.makeText(this, "Level 2 button clicked", Toast.LENGTH_SHORT).show()
        }

        btnLevel3.setOnClickListener {
            viewModel.startBuzzer(3)
            Toast.makeText(this, "Level 3 button clicked", Toast.LENGTH_SHORT).show()
        }

        btnLevel4.setOnClickListener {
            viewModel.startBuzzer(4)
            Toast.makeText(this, "Level 4 button clicked", Toast.LENGTH_SHORT).show()
        }


        // Observe StateFlow for sensor readings
        lifecycleScope.launch {
            viewModel.sensorReadings.collect { readings ->
                readings.forEach { (sensorId, reading) ->
                    updateUI(sensorId, reading)
                }
            }
        }

        Toast.makeText(this, "Ultrasonic monitoring started", Toast.LENGTH_SHORT).show()
    }

    private fun updateUI(sensorId: Int, reading: Int) {
        if (reading != -1) {


            val readingText = "$reading cm"

            when (sensorId) {
                0 -> txtUltrasonicOne.text = readingText
                1 -> txtUltrasonicTwo.text = readingText
                2 -> txtUltrasonicThree.text = readingText
                3 -> txtUltrasonicFour.text = readingText
                4 -> txtUltrasonicFive.text = readingText
                5 -> txtUltrasonicSix.text = readingText
                else -> Log.e("ULTRASONIC", "Invalid sensor ID: $sensorId")
            }
        }
    }
}
